### PR TITLE
Secure api

### PR DESCRIPTION
**Pull request by aidanouckama**
ADDED SECURITY TO BACKEND!

### CODE REVIEW

**Review by Nyumat — CHANGES_REQUESTED** [review]
Can we abstract fetch calls to a reusable method, so we don't need to repeat ourselves with this:

``` Authorization: `Bearer ${Cookies.get('sessionId')}`,```

For every fetch call in the client code?

What I'd do is make a class or method that idiomatically takes in a session ID and anonymous request data, then makes the corresponding fetch request with the authorization headers attached. DRY, baby!

Everything else LGTM.

**Comment by aidanouckama** [comment]
> Can we abstract fetch calls to a reusable method, so we don't need to repeat ourselves with this:
> 
> `` Authorization: `Bearer ${Cookies.get('sessionId')}`, ``
> 
> For every fetch call in the client code?
> 
> What I'd do is make a class or method that idiomatically takes in a session ID and anonymous request data, then makes the corresponding fetch request with the authorization headers attached. DRY, baby!
> 
> Everything else LGTM.

That's what I was thinking, definitely felt repetitive, I'll do that RN

**Comment by Nyumat** [comment]
This might add more code, but in the long run, I think we'll thank ourselves. Additionally, I'd look into [joi](https://www.npmjs.com/package/joi) for validating the schema (request body for each request)

**Comment by Nyumat** [comment]
> > Can we abstract fetch calls to a reusable method, so we don't need to repeat ourselves with this:
> > `` Authorization: `Bearer ${Cookies.get('sessionId')}`, ``
> > For every fetch call in the client code?
> > What I'd do is make a class or method that idiomatically takes in a session ID and anonymous request data, then makes the corresponding fetch request with the authorization headers attached. DRY, baby!
> > Everything else LGTM.
> 
> That's what I was thinking, definitely felt repetitive, I'll do that RN

Yeah..in the long run, I think we'll thank ourselves. Additionally, I'd look into [joi](https://www.npmjs.com/package/joi) for validating the schema (request body for each request)

File: client/src/context/SessionContext.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import { HOST } from '../util/apiRoutes'
+
+export const SessionContext = createContext()
+
+export const SessionContextProvider = ({ children }) => {
+    useEffect(() => {
+        // Function to initialize session
+        const initializeSession = async () => {
+            const currentSessionId = Cookies.get('sessionId')
+            console.log(currentSessionId)
```
Comment (arfpog) [inline client/src/context/SessionContext.js]: remove

File: client/src/context/SessionContext.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+            if (!currentSessionId || currentSessionId === 'undefined') {
+                // Use fetch to get a new session ID if one does not already exist
+                try {
+                    const response = await fetch(`${HOST}/api/token`)
+
+                    if (!response.ok) {
+                        throw new Error('Network response was not ok')
```
Comment (arfpog) [inline client/src/context/SessionContext.js]: what we want to do here is check if the session is expired. so if the token that we used before is too old, then we need to regenerate a token and send it here to be used.

File: server/middleware/token.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+  const token = tokenParts[1];
+
+  try {
+    jwt.verify(token, process.env.JWT_SECRET);
+    next();
+  } catch (err) {
+    console.error(err);
+    return res.status(400).json({ msg: "Invalid or expired token." });
```
Comment (arfpog) [inline server/middleware/token.js:36]: if token is expired, need to generate a new one

**Review by arfpog — CHANGES_REQUESTED** [review]

**Review by arfpog — APPROVED** [review]

**Review by Nyumat — APPROVED** [review]
marvelous. ggs.

**Pull request merged**